struct.md

```diff
@@ -0,0 +1,297 @@
+# Struct Proposal
+
+This document tries to define with examples what shall be (theoretically) supported in VSS after adding struct support.
+I.e. what you can do still claiming that the model is correct VSS.
+It only to a limited extent show implications for vss-tools, and then only for syntactic/semantic checks.
+As of now it does not state how exporters are affected.
+
+## Proposed VSS 4.0 acceptance criteria and increments
+
+We need to decide how far we must go before we can release VSS 4.0, we does not need to go the whole way in one step
+
+### Increment 1
+
+* Syntax can be used without VSS-tools complaining
+* No semantic check by VSS-tools
+* No documentation
+* Not used in VSS standard catalog
+* Struct not supported by VSS-tools exporters
+
+
+### Increment 2
+
+* Semantic check on type-references by VSS-tools
+* Syntax documented
+* Well defined behavior for all exporters (e.g. either support, and/or give warning if used)
+
+### Increment 3
+
+* All "standard" exporters in vss-tools support structs
+* Guidelines on when to consider using structs rather than branches documented
+
+### Increment 4
+
+* We may start to use structs in VSS standard catalog.
+
+
+## Simple Usage
+
+```
+DeliveryInfo:
+  type: struct
+  description: A struct type containing info for each delivery
+  
+DeliveryInfo.Address:
+  datatype: string
+  type: item
+  description: Destination address
+
+DeliveryInfo.Receiver:
+  datatype: string
+  type: item
+  description: Name of receiver
+
+DeliveryList:
+  datatype: DeliveryInfo
+  type: sensor
+  description: List of deliveries
+```
+
+For VSS 4.0 it is not necessary that vss-tools do semantic check, i.e. if someone would add an extra `f` by mistake like this:
+
+```
+
+DeliveryList:
+  datatype: DeliveryInffo
+  type: sensor
+  description: List of deliveries
+```
+
+... then VSS-tools does not necessarily need to give an error (stretch goal to have semantic check that referred type exist).
+
+## Name resolution
+
+For now, two ways of referring to a type shall be considered correct:
+
+### Reference to a struct definition within same branch
+
+As the example above.
+
+TBD: Or do we want a more flexible approach, i.e. if you specify "ABC" as datatype, that a tool shall search upwards in all parent branches?
+
+I.e. If a signal `A.B.C.D` is defined with type `X`, then the following priority order shall apply:
+
+* If `A.B.C.X` exists, then it will be used.
+* Else if `A.B.X` exists, then it will be used.
+* Else if `A.X` exists, then it will be used.
+
+### Reference by absolute path
+
+Reference by full path shall also be allowed. For now relative paths (e.g. `../Powertrain` shall not be supported).
+But vss-tools does not need to resolve or verify that type reference is correct in VSS 4.0.
+
+```
+
+DeliveryList:
+  datatype: Vehicle.Some.Branch.DeliveryInfo
+  type: sensor
+  description: List of deliveries
+```
+
+## Expectations on VSS implementations (e.g. VISS, KUKSA.val)
+
```

Comment:
Has this expectation been discussed in W3C, e.g. VISS? If not you have potential for divergence.

Reply:
No, not yet. And it is an interesting topic - how much should we think on or consider implications on downstream components/specifications. But I think it is a good topic to bring up within the W3C Automotive (and KUKSA.val) so that they at least are aware of the thoughts

Reply:
Created an issue at https://github.com/w3c/automotive/issues/461 to get their attention. 
FYI: @UlfBj @tguild @wonsuk73

Reply:
I just lost a long more detailed reply so I am just going to summarise...

Personally, I think there should be a parallel process that is using this 'big bang' change to capture input for Best Practices and Governance for two different cases, general (big) changes and this change specifically.

Example for general case is discussing with the other pillars.
Example for this change specifically is all the flags ppl have raised for impact on tooling and 'how not to misuse it'.

These in turn can also be useful input to the decision process.

> But I think it is a good topic to bring up within the W3C Automotive (and KUKSA.val) so that they at least are aware of the thoughts

With regards VISS and Kuksa.val I would separate them into two related categories, EG pillars and downstream eco-system which are consuming the deliverables.

VISS is part of the EG pillars and is standardised in a joint development agreement with W3C. So querying the feature/impact/roadmap with VISS would seem to be a natural process in the EG. That could (should) as equally apply to other pillars such as VSSo and likely (would need good argument for why not) lead to a formal step in the 'big bang' change processes in the EG.

None of that need be heavyweight process.

When Genivi had a formal compliance program for the IVI specification and Baselines to prove it was implementable such big changes would be incubated in the originating projects and normalised in the SAT/Baseline Integration Team (BIT). Then big changes, especially breaking changes, could be planned for. With something like CommonAPI that could have a wide affect it was very useful. It sometimes requires some pragmatic horse trading but it had a practical downstream affect for integration inside the companies firewalls and commercial products. We were often simultaneously agreeing both the roadmap and aligning production, which reduced integration headaches for participants. The latter encourages upstream participation.

I would put Kuksa.val in the downstream eco-system of projects consuming the EG deliverables. That includes the OSS VSS Data Servers such as Kuksa.val (VISS), WAII (VISS), Aos Edge (VISS) or the Covesa hosted in-vehicle GraphQL Server (GraphQL). Then of course there are other components both commercial and OSS such as AWS Fleetwise. Avoiding detailing the large subject of good community building I think you can just follow good OSS practise. Encourage people to be aware of and engage in upstream (don't complain to us if you don't), give good notice of 'big bangs' and solicit feedback. There is of course the link to Covesa marketing and things like announcement lists.

Reply:
I agree on that the general process better shall be defined. I think the process of this change so far has worked quite well, much of it started before the data expert group actually existed

- Started as a discussion in the project, discussed in several meetings and alternatives described in [Wiki](https://github.com/COVESA/vehicle_signal_specification/wiki/Complex-data-structures-in-VSS-based-systems)
- Discussed at a working sessions at an All Member Meeting, so it has been well announced to COVESA members.
- And now we have a proposal including proposed increments being discussed in VSS meetings
- Structs are also mentioned as a possible addition for VSS 4.0 in the roadmap

So if VISS/VSSo/KUKSA.val follows what happens in VSS it should not come as a surprise to them that struct support is coming. Then it is a different topic if or when they will support it. We already have concepts in VSS (like `aggregate`) that is ignored by many downstream implementations. I do not see it as an absolute must that e.g. VISS supports all concepts/signals of VSS, even if it is of course is preferable.

Reply:
Agreed that this particular change has been trailed for a long period.

It may have been clearer in the original longer post that I lost, but my point is that the actions carried out for this change can be captured and used as input into the best practices and governance tasks.

Further, that it is likely that non-functional requirements such as best practices are an aspect of this change, e.g. best practise explaining why you should not be suddenly abandoning branches in favours of structs - obviously the deliverable would be more nuanced. If the community agrees that to be the case, then there creation should be a first class requirement/task in the work breakdown for this feature.

Further still, in the process improvement task, there likely should be something (e.g checkbox) about formally asking pillars for feedback on big changes. A practitioner will have expertise from their perspective - 'I need this new features because of Foo'. The pillars may have other useful insights from other perspectives - security/ontology etc.

I was not thinking much on whether VISS or VSSo knows Foo support is coming or not. I was thinking more top down that the design of Foo takes the design goals of VSS and its associated standards into account. What impact Foo would have on VISS? Would anyone support it? Does Foo break the separation of concerns @danielwilms articulated for the over arching design of the VSS or does it inadvertently introduce issues? Would an ontology practitioner like @jdacoello see issues? I don't tag these people for them to answer here, but just to illustrate.

I agree that it might not be an absolute must that the related specs support all features of VSS. That said, wouldn't it be a general principal that they do where sensible. Even if it is later. If a potential adopter finds a complicated patch work of part support weaved through tooling and servers does that become a barrier to entry?

Reply:
@slawr I think the ontology version (i.e., `VSSo`) would not be affected much by the addition of `struct` in `VSS`. In fact, the ontology is much more expressive than a simple tree. It can contain multiple hierarchies that can be referenced to each other naturally using an intuitive graph representation. In this sense, the ontology could easily adapt to group entities. 

However, I see more problems in trying to define complex data structures with the `VSS` approach. `VSS` has been successfully used as a naming convention for univariate vehicle properties. Since there is only a top-down approach for the description of new items, one might need to cross-reference many times within the spec file when complex structures are needed. However, this doesn't mean that it is not possible. If we can at least collect 10 good examples of structs that are needed, then it makes total sense to try to serve them.

However, if these complex structures are about high-level information, then the name Vehicle _**Signal**_ Specification is miss leading. Signals happen at a lower level of abstraction than what VSS is actually covering at the moment.

Reply:
The discussion so far within VSS meeting have been that structs (in standard VSS catalog) only shall be used for data that logically belongs together. A hypothetical example could be that we in the future would like to have a signal `Vehicle.ADAS.ObstacleList` to present all obstacles currently identified by the vehicle, where obstacles are represented by a struct which may contain for example category/risk/probability. It shall not be used to group random data.